File: Week_03/51.n皇后.java
import java.util.Set;

/*
 * @lc app=leetcode.cn id=51 lang=java
 *
 * [51] N皇后
 */

// @lc code=start
class Solution {
    private int n;
    private List<List<String>> res;
    private Set <Integer> cols;
    private Set <Integer> main;
    private Set <Integer> vice;
    public List<List<String>> solveNQueens(int n) {
        this.n = n;
        res = new ArrayList<>();
        if (n <= 0) return res;
        cols = new HashSet<Integer>();
        main = new HashSet<Integer>();
        vice = new HashSet<Integer>();

        helper(0, new LinkedList<Integer> ());
        return res;

    }

    public void helper(int row, LinkedList<Integer> cur) {
        if (row == n) {
            List<String> sub_res = trans(cur);
            res.add(sub_res);
            return ;
        }

        for (int col = 0; col < n; col ++) {
            if (!cols.contains(col) && !main.contains(row + col) && !vice.contains(row - col)) {
                cur.add(col);
                cols.add(col);
                main.add(row + col);
                vice.add(row - col);

                helper(row + 1, cur);

                // reverse
                cur.remove(cur.size() - 1);
                cols.remove(col);
                main.remove(row + col);
                vice.remove(row - col);
            }
        }
        
    }

    public List<String> trans(LinkedList<Integer> cur) {
        List<String> res = new ArrayList<>();
        for (int num: cur) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < n; i ++) {
                sb.append(".");
            }
            sb.replace(num, num + 1, "Q");
            res.add(sb.toString());
        }
        return res;
    }
}
// @lc code=end

File: Week_05/51.n皇后.py
#
# @lc app=leetcode.cn id=51 lang=python3
#
# [51] N皇后
#

# @lc code=start
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        self.res = []
        self.n = n
        self.dfs(0, 0, 0, 0, []) 
        
        return [['.'*alp + 'Q' + '.' * (n-1-alp) for alp in sub_res] for sub_res in self.res]
        
    def dfs(self, row, cols, pie, na, cur_res):
        # terminator
        if self.n == row:
            self.res.append(cur_res)
            return
        
        # process current logic
        bits = (~(cols | pie | na)) & ((1 << self.n) - 1)
        while bits:
            pos = bits & -bits # 保留最低位1
            # 计算存放的列
            tmp = pos
            cnt = 0
            while tmp & 1 == 0:
                tmp >>= 1
                cnt += 1
            bits &= (bits - 1) # 将最低位1为清零
            
            # drill down
            self.dfs(row+1, cols | pos, (pie | pos) << 1, (na | pos) >> 1, cur_res + [cnt])
        
    
                
# @lc code=end

